starknet/l2_minter/src/ethereum_minter.cairo

```diff
@@ -0,0 +1,252 @@
+            assert(self.sold_out(), 'Contract not sold out');
+
+            // [Check] Booking ok;
+            let mut booking = self._booked_values.read((user_address.into(), id));
+            assert(
+                booking.status == mint_status_to_u8(MintStatus::Booked(())), 'Booking not found'
+            );
+
+            let projects_contract = self._projects_contract.read();
+            let slot = self._slot.read();
+
+            // [Interaction] Mint
+            let token_id = self
+                ._projects_contract
+                .read()
+                .mintNew(user_address.into(), slot, booking.value);
+
+            // [Effect] Update Booking status
+            booking.status = mint_status_to_u8(MintStatus::Minted(()));
+            self._booked_values.write((user_address.into(), id), booking);
+
+            // [Effect] Emit event
+            self
+                .emit(
+                    Event::BookingClaimed(
+                        BookingClaimed { address: user_address, value: booking.value,  }
+                    )
+                );
+        }
+
+        fn set_l1_minter_address(ref self: ContractState, l1_address: felt252) {
+            assert(!l1_address.is_zero(), 'L1 address cannot be zero');
+            let _l1_address = self._l1_minter_address.read();
+            assert(_l1_address.is_zero(), 'L1 address already set');
+            self._l1_minter_address.write(l1_address);
+        }
+
+        fn upgrade(ref self: ContractState, impl_hash: ClassHash) {
+            assert(!impl_hash.is_zero(), 'Class hash cannot be zero');
+            starknet::replace_class_syscall(impl_hash).unwrap_syscall();
+            self.emit(Event::Upgraded(Upgraded { implementation: impl_hash }));
+        }
+    }
+
+    #[l1_handler]
+    // TODO: Add L1 user address?
+    fn book_value_from_l1(
+        ref self: ContractState,
+        from_address: felt252,
+        user_address: ContractAddress,
+        value: u256, // TODO: u128 enough?
+        amount: u256, // TODO: u128 enough?
+        time: u64, // TODO: timestamp from L1?
+    ) {
+        // Can only be called by L1 minter
+        // This method shouldn't fail otherwise.
+        assert(from_address == self._l1_minter_address.read(), 'Only L1 minter can mint value');
+
+        // Get Booking Status
+        let new_user_mint_id = self._l1_mint_counts.read(user_address) + 1_u32;
+        self._l1_mint_counts.write(user_address, new_user_mint_id);
+        let unit_price = self._unit_price.read();
+        let current_max_supply = self._current_max_supply.read();
+        let max_value_per_tx = self._max_value_per_tx.read();
+        let min_value_per_tx = self._min_value_per_tx.read();
+
+        let mut status = MintStatus::Failed(());
+        if (value <= max_value_per_tx && value >= min_value_per_tx && amount == unit_price
+            * value && value <= current_max_supply) {
```

Comment:
If `current_max_supply` stands for `remaining_supply` we should consider to rename it.

Reply:
done